ambry-frontend/src/main/java/com/github/ambry/frontend/AmbrySecurityService.java

```diff
@@ -234,6 +240,17 @@ public void processResponse(RestRequest restRequest, RestResponseChannel respons
         exception = e;
       }
     }
+    if (quotaConfig.requestQuotaThrottlingEnabled) {
+      if (quotaConfig.quotaThrottlingMode == QuotaMode.THROTTLING) {
+        // TODO use enforcement recommendation to create more specific error message and set specific response headers.
+        // TODO make sure that container and account both are populated in request here
+        quotaManager.shouldThrottleOnRequestAndCharge(restRequest, blobInfo, new ArrayList<>());
```

Comment:
For Ambry, the actual cost of a request will be known only when the request has actually been processed (blob has been read or written). So before processing the request we check the current values of quota usage, to determine if the quota limits have already been crossed even before processing the request. At that point AsyncRequestResponseHandler::handleRequest(), we also check to throttle is the quota limit is crossed.

Here, in AmbrySecurityService, we call the shouldThrottleAndCharge method to actually charge for the request after we have processed the request. Since the request is already processed, there is no point of throttling here.

Reply:
I have renamed method to `charge`  instead of `shouldThrottleAndCharge` to better reflect the use.